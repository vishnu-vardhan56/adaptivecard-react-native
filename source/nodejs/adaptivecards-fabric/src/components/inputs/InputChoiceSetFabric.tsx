// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import * as React from "react";
import * as AC from "adaptivecards";
import * as FabricUI from "@fluentui/react";
import * as Shared from "../../utils/shared";

export class InputChoiceSetFabric extends Shared.ReactInputElement {

    private isMultiSelect: boolean;
    private choices: AC.Choice[];
    private style: string;
    private selectedValues: string[] = [];

    public parse = (json: any, errors?: AC.IValidationError[]) => {
        super.parse(json, errors);
        this.value = this.defaultValue;
        this.selectedValues = this.defaultValueToArray(this.value);
        this.isMultiSelect = AC.getBoolValue(json.isMultiSelect, false);
        this.title = AC.getStringValue(json.title);
        this.style = AC.getStringValue(json.style);
        this.parseChoises(json);
    }

    protected renderReact = (): JSX.Element => (
        this.isCompact() ?
            <this.ComboBox /> :
            this.createExpandedView()
    )

    public getJsonTypeName = (): string => "Input.ChoiceSet";

    public toJSON = () => {
        let result = super.toJSON();

        if (this.choices.length > 0) {
            AC.setProperty(result, "choices", this.choices.map(ch => ch.toJSON()));
        }

        if (!this.isCompact()) {
            AC.setProperty(result, "style", "expanded", false);
        }

        AC.setProperty(result, "isMultiSelect", this.isMultiSelect, false);

        return result;
    }

    private parseChoises = (json: any) => {
        this.choices = Array.isArray(json.choices) ?
            json.choices.map(ch => {
                let choice = new AC.Choice();
                choice.parse(ch);
                return choice;
            }) :
            [];
    }

    private isCompact = () => this.style === "compact";

    private ComboBox = (): JSX.Element => (
        <FabricUI.ComboBox
            onChange={this.handleComboBoxChange}
            id={this.id}
            options={this.inputChoicesToComboBoxOptions(this.choices)}
            multiSelect={this.isMultiSelect}
            selectedKey={this.defaultValueToArray(this.value)}
        />
    )

    private createExpandedView = (): JSX.Element => {
        return this.isMultiSelect ?
            this.inputChoicesToCheckbox() :
            this.createChoiceGroup();
    }

    private createChoiceGroup = (): JSX.Element => (
        <FabricUI.ChoiceGroup
            id={this.id}
            onChange={this.handleChoiceGroupChange}
            options={this.inputChoicesToChoiceGroupOptions(
                this.choices,
                this.defaultValueToArray(this.value))}
            multiple={this.isMultiSelect}
        />
    )

    private inputChoicesToCheckbox = (): JSX.Element => (
        <React.Fragment>
            {this.choices.map((c, i) =>
                <FabricUI.Checkbox
                    key={c.value}
                    label={c.title}
                    defaultChecked={this.defaultValueToArray(this.value).includes(c.value)}
                    onChange={(ev, checked) => {
                        this.updateMultiselectData(checked, c.value);
                        ev.stopPropagation();
                        ev.preventDefault();
                    }}
                    styles={{
                        root: {
                            paddingTop: i > 0 ? this.hostConfig.spacing.default : 0,
                        },
                    }}
                />,
            )}
        </React.Fragment>
    )

    private defaultValueToArray = (value: string): string[] => {
        if (value) {
            return value.split(",");
        }
        return [];
    }

    private inputChoicesToComboBoxOptions = (choices: AC.Choice[]): FabricUI.IComboBoxOption[] => {
        return choices.map<FabricUI.IComboBoxOption>(c => ({ key: c.value, text: c.title }));
    }

    private inputChoicesToChoiceGroupOptions = (choices: AC.Choice[], value: string[]): FabricUI.IChoiceGroupOption[] => {
        return choices.map<FabricUI.IChoiceGroupOption>(c => ({ key: c.value, text: c.title, checked: value.includes(c.value) }));
    }

    private handleChoiceGroupChange =
        (ev?: React.FormEvent<HTMLElement | HTMLInputElement>, option?: FabricUI.IChoiceGroupOption): void => {
            this.value = option.key;
            ev.stopPropagation();
            ev.preventDefault();
        }

    private handleComboBoxChange =
        (event: React.FormEvent<FabricUI.IComboBox>, option?: FabricUI.IComboBoxOption, index?: number, value?: string): void => {
            if (this.isMultiSelect) {
                this.updateMultiselectData(option.selected, option.key);
            } else {
                this.value = `${option.key}`;
            }
            event.stopPropagation();
            event.preventDefault();
        }

    private updateMultiselectData = (selected: boolean, key: any): void => {
        this.updateSelectedValues(selected, key);
        this.value = this.selectedValues.join(",");
    }

    private updateSelectedValues = (selected: boolean, key: any): void => {
        selected ? this.selectedValues.push(key) : this.removeItemFromArray(this.selectedValues, key);
    }

    private removeItemFromArray = (arr: any[], item: any): void => {
        const index = arr.indexOf(item);
        arr.splice(index, 1);
    }
}
