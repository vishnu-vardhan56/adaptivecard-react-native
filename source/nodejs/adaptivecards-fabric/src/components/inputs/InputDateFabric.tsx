// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import * as React from "react";
import * as FabricUI from "@fluentui/react";
import * as AC from "adaptivecards";
import * as Shared from "../../utils/shared";

export class InputDateFabric extends Shared.ReactInputElement {

    private date: Date;
    private placeholder: string;
    private minDate: Date;
    private maxDate: Date;

    protected renderReact = (): JSX.Element => {
        return this.buildDatePicker();
    }

    public getJsonTypeName = (): string => "Input.Date";

    public parse = (json: any, errors?: AC.IValidationError[]) => {
        super.parse(json, errors);
        this.placeholder = AC.getStringValue(json.placeholder);
        this.parseDates(json, errors);
    }

    public toJSON = () => {
        let result = super.toJSON();

        AC.setProperty(result, "min", this.minDate);
        AC.setProperty(result, "max", this.maxDate);
        AC.setProperty(result, "placeholder", this.placeholder);

        return result;
    }

    private parseDates = (json: any, errors?: AC.IValidationError[]) => {
        this.value = this.defaultValue;
        this.date = this.value ? this.getDate(this.value) : new Date();
        this.minDate = json.min ? this.getDate(AC.getStringValue(json.min)) : undefined;
        this.maxDate = json.max ? this.getDate(AC.getStringValue(json.max)) : undefined;
    }

    private getDate = (dateString: string, errors?: AC.IValidationError[]): Date => {
        try {
            return new Date(dateString);
        } catch (error) {
            Shared.raiseParseError({
                error: AC.ValidationError.InvalidPropertyValue,
                message: error,
            }, errors);
            return undefined;
        }
    }

    private buildDatePicker = () => {
        return (<FabricUI.DatePicker
            id={this.id}
            onSelectDate={this.handleSelectDate}
            value={this.date}
            placeholder={this.placeholder}
            allowTextInput={true}
            minDate={this.minDate}
            maxDate={this.maxDate}
        />);
    }

    private handleSelectDate = (date: Date) => {
        this.value = date.toLocaleDateString("en-US");
    }
}
