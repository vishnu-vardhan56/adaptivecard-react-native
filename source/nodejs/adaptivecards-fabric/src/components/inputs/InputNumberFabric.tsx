// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import * as React from "react";
import * as AC from "adaptivecards";
import * as FabricUI from "@fluentui/react";
import * as Shared from "../../utils/shared";

export class InputNumberFabric extends Shared.ReactInputElement {

    private max?: number;
    private min?: number;
    private placeholder: string;

    public parse = (json: any, errors?: AC.IValidationError[]) => {
        super.parse(json, errors);
        this.value = this.defaultValue;
        this.min = Shared.getIntValue(json.min);
        this.max = Shared.getIntValue(json.max);
        this.placeholder = AC.getStringValue(json.placeholder);
    }

    protected renderReact = (): JSX.Element => (
        <FabricUI.TextField
            id={this.id}
            placeholder={`${this.placeholder}`}
            value={`${this.defaultValue}`}
            type="number"
            max={this.max}
            min={this.min}
            onChange={this.handleChange}
        />
    )

    public getJsonTypeName = (): string => "Input.Number";

    public toJSON = () => {
        let result = super.toJSON();

        AC.setProperty(result, "placeholder", this.placeholder);
        AC.setProperty(result, "min", this.min);
        AC.setProperty(result, "max", this.max);

        return result;
    }
}
