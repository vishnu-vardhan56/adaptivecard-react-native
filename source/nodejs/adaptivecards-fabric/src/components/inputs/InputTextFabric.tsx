// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import * as React from "react";
import { TextField } from "@fluentui/react";
import * as AC from "adaptivecards";
import * as Shared from "../../utils/shared";

export class InputTextFabric extends Shared.ReactInputElement {

    private isMultiline: boolean;
    private maxLength: number;
    private placeholder: string;
    private label: string;
    private style: AC.InputTextStyle = AC.InputTextStyle.Text;

    protected renderReact = (): JSX.Element => {
        return this.buildInput();
    }

    public getJsonTypeName = (): string => "Input.Text";

    public parse = (json: any, errors?: any) => {
        super.parse(json, errors);
        this.value = this.defaultValue;
        this.isMultiline = AC.getBoolValue(json.isMultiline, false);
        this.maxLength = Shared.getIntValue(json.maxLength);
        this.placeholder = AC.getStringValue(json.placeholder);
        this.label = AC.getStringValue(json.label);
        this.style = AC.getEnumValue(AC.InputTextStyle, json.style, AC.InputTextStyle.Text);
        this.inlineAction = AC.createActionInstance(this, json.inlineAction, [AC.ShowCardAction.JsonTypeName], false, errors);
    }

    public toJSON = () => {
        let result = super.toJSON();

        AC.setProperty(result, "placeholder", this.placeholder);
        AC.setProperty(result, "maxLength", this.maxLength, 0);
        AC.setProperty(result, "isMultiline", this.isMultiline, false);
        AC.setEnumProperty(AC.InputTextStyle, result, "style", this.style, AC.InputTextStyle.Text);
        if (this.inlineAction) {
            AC.setProperty(result, "inlineAction", this.inlineAction.toJSON());
        }

        return result;
    }

    private buildInput = (): JSX.Element => {
        return (
            <TextField
                id={this.id}
                type={AC.InputTextStyle[this.style].toLowerCase()}
                value={this.defaultValue}
                multiline={this.isMultiline}
                maxLength={this.maxLength}
                placeholder={this.placeholder}
                label={this.label}
                onChange={this.handleChange}
                onKeyDown={this.handleKeyDown}
            />
        );
    }

    private handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement | HTMLTextAreaElement>): void => {
        // Enter pressed
        if (e.keyCode === 13 && this.inlineAction) {
            this.inlineAction.execute();
        }
    }
}
