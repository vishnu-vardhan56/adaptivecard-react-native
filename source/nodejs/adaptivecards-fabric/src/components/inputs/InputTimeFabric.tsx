// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import * as React from "react";
import * as AC from "adaptivecards";
import * as FabricUI from "@fluentui/react";
import * as Shared from "../../utils/shared";

export class InputTimeFabric extends Shared.ReactInputElement {

    private placeholder: string;
    private min: string;
    private max: string;

    public parse = (json: any, errors?: AC.IValidationError[]) => {
        super.parse(json, errors);
        this.value = this.defaultValue;
        this.placeholder = AC.getStringValue(json.placeholder);
        this.min = AC.getStringValue(json.min);
        this.max = AC.getStringValue(json.max);
    }

    protected renderReact = (): JSX.Element => {
        return (
            <FabricUI.TextField
                id={this.id}
                placeholder={`${this.placeholder}`}
                value={`${this.defaultValue}`}
                type="time"
                onChange={this.handleChange}
                max={this.max}
                min={this.min}
            />
        );
    }

    public getJsonTypeName = (): string => "Input.Time";

    public toJSON = () => {
        let result = super.toJSON();

        AC.setProperty(result, "min", this.min);
        AC.setProperty(result, "max", this.max);

        return result;
    }
}
