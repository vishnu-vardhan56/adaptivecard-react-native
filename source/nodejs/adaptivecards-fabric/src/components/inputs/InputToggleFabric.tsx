// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import * as React from "react";
import * as AC from "adaptivecards";
import * as FabricUI from "@fluentui/react";
import * as Shared from "../../utils/shared";

export class InputToggleFabric extends Shared.ReactInputElement {

    private valueOn: string;
    private valueOff: string;

    public parse = (json: any, errors?: AC.IValidationError[]) => {
        super.parse(json, errors);
        this.value = this.defaultValue;
        this.valueOn = AC.getStringValue(json.valueOn);
        this.valueOff = AC.getStringValue(json.valueOff);
        this.title = AC.getStringValue(json.title);
    }

    protected renderReact = (): JSX.Element => (
        <FabricUI.Toggle
            id={this.id}
            inlineLabel={true}
            onChange={this.handleToggleChange}
            defaultChecked={this.defaultValue === this.valueOn}
            label={this.title}
            styles={{
                root: {
                    marginBottom: 0,
                    marginTop: 0,
                },
            }}
        />
    )

    public getJsonTypeName = (): string => "Input.Toggle";

    public toJSON = () => {
        let result = super.toJSON();

        AC.setProperty(result, "valueOn", this.valueOn, "true");
        AC.setProperty(result, "valueOff", this.valueOff, "false");

        return result;
    }

    private handleToggleChange = (event: React.MouseEvent<HTMLElement> | React.ChangeEvent, checked?: boolean) => {
        this.value = checked ? this.valueOn : this.valueOff;
    }
}
