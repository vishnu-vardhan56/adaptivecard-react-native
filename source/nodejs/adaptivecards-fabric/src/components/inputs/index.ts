export * from "./InputChoiceSetFabric";
export * from "./InputDateFabric";
export * from "./InputNumberFabric";
export * from "./InputTextFabric";
export * from "./InputTimeFabric";
export * from "./InputToggleFabric";
