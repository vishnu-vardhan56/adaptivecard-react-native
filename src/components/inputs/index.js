export * from './text-input';
export * from './number-input';
export * from './toggle-input';
export * from './date-input';
export * from './time-input';
export * from './picker-input';
export * from './choice-set/choice-set-input';