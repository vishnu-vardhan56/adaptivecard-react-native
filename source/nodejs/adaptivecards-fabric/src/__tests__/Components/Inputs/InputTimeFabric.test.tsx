// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { InputTimeFabric } from "../../../components/inputs/InputTimeFabric";
import * as TestUtils from "../../TestUtils";
import { initializeIcons } from "@fluentui/react";

initializeIcons();

describe("InputTimeExtended", () => {
    it("should be rendered successfully", () => {
        let input = new InputTimeFabric();

        expect(input).toBeDefined();
        expect(input.render()).toBeDefined();
    });

    it("should render input time correctly", () => {
        let input = TestUtils.renderFabricComponent(InputTimeFabric, {
            "type": "Input.Time",
            "id": "time",
            "min": "09:00",
            "max": "17:00",
            "value": "15:30",
        });

        expect(input).toBeDefined();
        expect(input).toMatchSnapshot();
    });

    it("can get json name", () => {
        expect(new InputTimeFabric().getJsonTypeName()).toBe("Input.Time");
    });

    it("generates json correctly", () => {
        const element = new InputTimeFabric();
        element.parse({
            "type": "Input.Time",
            "id": "time",
            "min": "09:00",
            "max": "17:00",
            "value": "15:30",
        });

        const json = element.toJSON();
        expect(json).toBeDefined();
        expect(json.type).toEqual("Input.Time");
        expect(json.value).toEqual("15:30");
        expect(json.max).toEqual("17:00"); 
        expect(json.min).toEqual("09:00");

    });
});
