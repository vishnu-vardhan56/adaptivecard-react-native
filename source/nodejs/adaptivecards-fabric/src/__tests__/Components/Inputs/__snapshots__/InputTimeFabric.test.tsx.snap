// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputTimeExtended should render input time correctly 1`] = `
<div
  style="display: flex; flex-direction: column; box-sizing: border-box; flex: 0 0 auto;"
>
  <div
    class="ac-input-container"
    style="display: flex;"
  >
    <div
      style="width: 100%; min-width: 0px;"
    >
      <div
        class="ms-TextField root-41"
      >
        <div
          class="ms-TextField-wrapper"
        >
          <div
            class="ms-TextField-fieldGroup fieldGroup-42"
          >
            <input
              aria-invalid="false"
              class="ms-TextField-field field-43"
              id="time"
              max="17:00"
              min="09:00"
              placeholder="undefined"
              type="time"
              value="15:30"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
